File: lenskit-archetype-simple-analysis/src/main/resources/archetype-resources/chart.R
# Create a chart comparing the algorithms

library("ggplot2")
library("grid")

all.data <- read.csv("eval-results.csv")

all.agg <- aggregate(cbind(RMSE.ByRating, RMSE.ByUser, nDCG)
                     ~ Algorithm,
                     data=all.data, mean)

rmse.both <- rbind(
  data.frame(Algorithm=all.agg$Algorithm, RMSE=all.agg$RMSE.ByRating,
             mode="Global"),
  data.frame(Algorithm=all.agg$Algorithm, RMSE=all.agg$RMSE.ByUser,
             mode="Per-User"))

chart.rmse <- qplot(RMSE, Algorithm, data=rmse.both, shape=mode, color=mode)

chart.ndcg <- qplot(nDCG, Algorithm, data=all.agg)

chart.build <- ggplot(all.data, aes(Algorithm, BuildTime / 1000)) +
  geom_boxplot() +
  ylab("Build time (seconds)")

chart.test <- ggplot(all.data, aes(Algorithm, TestTime / 1000)) +
  geom_boxplot() +
  ylab("Test time (seconds)")

print("Outputting to accuracy.pdf")
pdf("accuracy.pdf", paper="letter", width=0, height=0)
error.layout <- grid.layout(nrow=2, heights=unit(0.5, "npc"))
pushViewport(viewport(layout=error.layout, layout.pos.col=1))
print(chart.rmse, vp=viewport(layout.pos.row=1))
print(chart.ndcg, vp=viewport(layout.pos.row=2))
popViewport()
dev.off()

print("Outputting to speed.pdf")
pdf("speed.pdf", paper="letter", width=0, height=0)
times.layout <- grid.layout(nrow=2, heights=unit(0.5, "npc"))
pushViewport(viewport(layout=times.layout, layout.pos.col=1))
print(chart.build, vp=viewport(layout.pos.row=1))
print(chart.test, vp=viewport(layout.pos.row=2))
popViewport()
dev.off()

File: lenskit-archetype-simple-analysis/src/main/resources/archetype-resources/chart.py
#!/usr/bin/env python

# This script generates a plot of the results.
# It requires Pandas, the Python data table library: http://pandas.pydata.org

import pandas as pd
import numpy as np
import matplotlib
matplotlib.use('svg')
import matplotlib.pyplot as plt

# Import data
frame = pd.read_csv('eval-results.csv')

# Compute means of metrics
means = frame.groupby('Algorithm').mean()

xvals = np.arange(len(means))

# Plot the RMSE
plots = plt.plot(xvals,
                 means['RMSE.ByUser'], 'ro',
                 means['RMSE.ByRating'], 'bs')
plt.legend(('By-User', 'Global'), loc='lower right', numpoints=1)
plt.xlim(-0.5, len(means) - 0.5)
plt.xticks(xvals, means.index)
plt.xlabel('Algorithm')
plt.ylabel('RMSE')
plt.savefig('rmse.svg')

# Plot the nDCG
plt.clf()
plots = plt.plot(xvals, means['nDCG'], 'ro')
plt.xlim(-0.5, len(means) - 0.5)
plt.xticks(xvals, means.index)
plt.xlabel('Algorithm')
plt.ylabel('nDCG')
plt.savefig('ndcg.svg')

# Plot the build and test time
plt.clf()
frame.boxplot('BuildTime', 'Algorithm')
plt.savefig('build-time.svg')

plt.clf()
frame.boxplot('TestTime', 'Algorithm')
plt.savefig('test-time.svg')
